homeassistant/components/reolink/config_flow.py

```
@@ -0,0 +1,137 @@
+                errors[CONF_HOST] = "cannot_connect"
+            except CredentialsInvalidError:
+                errors[CONF_HOST] = "invalid_auth"
+            except ApiError:
+                errors[CONF_HOST] = "cannot_connect"
+            except Exception:  # pylint: disable=broad-except
+                _LOGGER.exception("Unexpected exception")
+                errors[CONF_HOST] = "unknown"
+
+            if not errors:
+                user_input[CONF_PORT] = self.port
+                user_input[CONF_USE_HTTPS] = self.use_https
+
+                await self.async_set_unique_id(self.unique_id, raise_on_progress=False)
+                self._abort_if_unique_id_configured(updates=user_input)
+
+                return self.async_create_entry(
+                    title=str(self.host_name), data=user_input
+                )
+
+        data_schema = vol.Schema(
+            {
+                vol.Required(CONF_USERNAME, default="admin"): str,
+                vol.Required(CONF_PASSWORD): str,
+                vol.Required(CONF_HOST): str,
+            }
+        )
+        if errors:
+            data_schema = data_schema.extend(
+                {
+                    vol.Optional(CONF_PORT): cv.positive_int,
+                    vol.Optional(CONF_USE_HTTPS): bool,
+                }
+            )
+
+        return self.async_show_form(
+            step_id="user", data_schema=data_schema, errors=errors
+        )
+
+    async def async_obtain_host_settings(
+        self, hass: core.HomeAssistant, user_input: dict
+    ):
+        """Initialize the Reolink host and get the host information."""
+        host = ReolinkHost(hass, user_input, {})
+
+        try:
+            if not await host.init():
+                _LOGGER.error(
+                    "Error while performing initial setup of %s:%i: failed to obtain data from device",
+                    host.api.host,
+                    host.api.port,
+                )
+                raise CannotConnect
```

You could raise the exception and use placeholders to propagate the error to the UI instead

example:
https://github.com/home-assistant/core/blob/6234190cfe3658a7f417c845b917f0f082c233ce/homeassistant/components/homekit_controller/config_flow.py#L483

---

Thanks for the feedback, 
I have split out cannot_connect and api_error, and added the error message as a placeholder to be shown in the UI directly